File: tech-studio-projects/financial-industry/crypto-trading-simulation/java/tests/performance_tests/performance_tests.cpp
#include <gtest/gtest.h>
#include <chrono>
#include "core/market_maker/MarketMaker.h"
#include "core/order_router/OrderRouter.h"
#include "core/data_feeds/MarketDataFeed.h"

// Helper function to simulate order routing time
void simulateOrderRouting(OrderRouter& orderRouter) {
    // Simulate some work for order routing
    for (int i = 0; i < 10000; ++i) {
        orderRouter.routeOrder(); // Call routeOrder multiple times
    }
}

// Performance Test for OrderRouter
TEST(OrderRouterPerformanceTest, RoutingPerformance) {
    OrderRouter orderRouter;
    orderRouter.start(); // Start the order router

    auto start = std::chrono::high_resolution_clock::now(); // Start time measurement
    simulateOrderRouting(orderRouter); // Simulate order routing
    auto end = std::chrono::high_resolution_clock::now(); // End time measurement

    std::chrono::duration<double, std::milli> duration = end - start; // Calculate duration
    std::cout << "Order routing took: " << duration.count() << " ms" << std::endl;

    orderRouter.stop(); // Stop the order router
    EXPECT_LE(duration.count(), 500); // Expect routing to take less than 500ms
}

// Performance Test for MarketMaker
TEST(MarketMakerPerformanceTest, MarketMakingPerformance) {
    MarketMaker marketMaker;
    marketMaker.start(); // Start the market maker

    auto start = std::chrono::high_resolution_clock::now(); // Start time measurement
    for (int i = 0; i < 1000; ++i) {
        // Simulate market making logic
        marketMaker.start(); // This should represent some market-making logic
        marketMaker.stop();  // This should represent stopping the market-making logic
    }
    auto end = std::chrono::high_resolution_clock::now(); // End time measurement

    std::chrono::duration<double, std::milli> duration = end - start; // Calculate duration
    std::cout << "Market making took: " << duration.count() << " ms" << std::endl;

    EXPECT_LE(duration.count(), 1000); // Expect total market-making to take less than 1000ms
    marketMaker.stop(); // Ensure market maker is stopped
}

File: tech-studio-projects/financial-industry/cryptocurrency-trading-simulation/java/tests/performance_tests/performance_tests.java
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertTrue;

// Helper method to simulate order routing time
void simulateOrderRouting(OrderRouter orderRouter) {
    // Simulate some work for order routing
    for (int i = 0; i < 10000; i++) {
        orderRouter.routeOrder(); // Call routeOrder multiple times
    }
}

public class PerformanceTests {

    private OrderRouter orderRouter;
    private MarketMaker marketMaker;

    @BeforeEach
    public void setUp() {
        orderRouter = new OrderRouter();
        marketMaker = new MarketMaker();
        orderRouter.start(); // Start the order router
        marketMaker.start(); // Start the market maker
    }

    @AfterEach
    public void tearDown() {
        orderRouter.stop(); // Stop the order router
        marketMaker.stop(); // Stop the market maker
    }

    @Test
    public void routingPerformanceTest() {
        long start = System.nanoTime(); // Start time measurement
        simulateOrderRouting(orderRouter); // Simulate order routing
        long end = System.nanoTime(); // End time measurement

        long duration = (end - start) / 1_000_000; // Calculate duration in milliseconds
        System.out.println("Order routing took: " + duration + " ms");

        assertTrue(duration <= 500, "Expected routing to take less than 500ms");
    }

    @Test
    public void marketMakingPerformanceTest() {
        long start = System.nanoTime(); // Start time measurement
        for (int i = 0; i < 1000; i++) {
            // Simulate market making logic
            marketMaker.start(); // This should represent some market-making logic
            marketMaker.stop();  // This should represent stopping the market-making logic
        }
        long end = System.nanoTime(); // End time measurement

        long duration = (end - start) / 1_000_000; // Calculate duration in milliseconds
        System.out.println("Market making took: " + duration + " ms");

        assertTrue(duration <= 1000, "Expected total market-making to take less than 1000ms");
    }

    public static void main(String[] args) {
        org.junit.runner.JUnitCore.main("PerformanceTests");
    }
}
